river/tree/_nodes/sgt_nodes.py

```diff
@@ -0,0 +1,238 @@
+import math
+import numbers
+import sys
+from typing import Dict, Hashable, Optional, Union
+
+from river import stats
+from river.base.typing import FeatureName, Target
+
+from .._utils import FeatureQuantizer, GradHess, GradHessStats
+
+
+class SGTSplit:
+    """ Split Class of the Streaming Gradient Trees (SGT).
+
+    SGTs only have one type of node. Their nodes, however, can carry a split object that indicates
+    they have been split and are not leaves anymore.
+    """
+
+    def __init__(
+        self,
+        feature_idx: FeatureName = None,
+        feature_val: Target = None,
+        is_nominal=False,
+    ):
+        # loss_mean and loss_var are actually statistics that approximate the *change* in loss.
+        self.loss_mean = 0.0
+        self.loss_var = 0.0
+        self.delta_pred = None
+
+        self.feature_idx = feature_idx
+        self.feature_val = feature_val
+        self.is_nominal = is_nominal
+
+
+class SGTLearningNode:
+    def __init__(self, prediction=0.0, depth=0):
+        self._prediction = prediction
+        self.depth = depth
+        self.last_split_attempt_at = 0
+
+        # Split test
+        self._split: Optional[SGTSplit] = None
+        self._children: Optional[Dict[Hashable, "SGTLearningNode"]] = None
+        self._split_stats: Optional[
+            Dict[FeatureName, Union[Dict[Hashable, GradHessStats], FeatureQuantizer]]
+        ] = {}
+        self._update_stats = GradHessStats()
+
+    def reset(self):
+        self._split_stats = {}
+        self._update_stats = GradHessStats()
+
+    def sort_instance_to_leaf(self, x) -> "SGTLearningNode":
```

Comment:
Does river deal with missing values explicitly, or is it up to the user to do a feature transformation to get rid of the problem? The original Moa code for SGTs has a check for missing values (represented as -1) which is not included here.

Reply:
Hi @henrygouk, we take advantage of `river`'s sparse dictionary-based inputs and simply ignore the missing features from the instances. On the other hand, features with `None` as their value might break our models.

The user is also free to use pre-processing strategies, such as those implemented in the `compose` module.

We have a [set of tests](https://github.com/online-ml/river/blob/00144d33d0c8c67c08dd9436edc89ee22d84f458/river/utils/estimator_checks.py#L154-L173) that ensure the models don't break if new features appear or features are missing from incoming instances.

In the case of this implementation of SGT, if a decision feature is missing from an instance, the tree sorts the data to the most "traversed" branch. This is also done in Hoeffding Trees.

Reply:
You are correct :)

When a new category arises, we create a new branch for it and sort the instance to this fresh leaf.  That is also done in the Hoeffding Trees. What do you think of this approach?

Reply:
Maybe we could provide more details in the docstring.